front/lib/api/assistant/actions/visualization.ts

```diff
@@ -175,16 +180,61 @@ export class VisualizationConfigurationServerRunner extends BaseActionConfigurat
       }),
     };
 
-    // Turn the conversation into a digest that can be presented to the model.
+    const contentFragmentsText = await Promise.all(
+      conversation.content
+        .flat(1)
+        .filter((m) => isContentFragmentType(m))
+        .map(async (m) => {
+          // This is needed because the ContentFragmentType is not actually narrowed correctly,
+          // this type is actually pretty complex and even if VSCode figure it out, `npm run build` does not.
+          // @flavien knows about this.
+          assert(isContentFragmentType(m));
+          if (!m.fileId) {
+            return;
+          }
+          if (m.contentType.startsWith("text/")) {
+            const file = await FileResource.fetchById(auth, m.fileId);
+            if (file === null) {
+              return null;
+            }
+            const readStream = file.getReadStream({
+              auth,
+              version: "original",
+            });
+            const lines = await readFirstFiveLines(readStream);
+            return lines;
+          } else {
+            return [];
+          }
+        })
+    );
+
+    // The prompt is a list of files that the visualization action can access.
+    // I have empirically found that this format works well for the model to generate the appropriate code,
+    // but feel free to change it if you think otherwise.
+    const prompt =
+      `You have access to the following files:\n` +
+      conversation.content
+        .flat(1)
+        .filter((m) => isContentFragmentType(m))
+        .map((m, i) => {
+          // This is needed because the ContentFragmentType is not actually narrowed correctly,
+          // this type is actually pretty complex and even if VSCode figure it out, `npm run build` does not.
+          // @flavien knows about this.
+          assert(isContentFragmentType(m));
```

Comment:
Same let's remove this and use an `as`

Reply:
if we don't want an as we could remove the filter above, return null if !isContentFragmentType and use the `removeNulls` utility to get rid of all the nulls we returned (and properly narrow the type)

Reply:
Will be removed in a follow up PR.